community/kernel/src/main/java/org/neo4j/kernel/impl/api/KernelStatement.java

```diff
@@ -58,6 +59,11 @@ public KernelStatement( KernelTransactionImplementation transaction,
     @Override
     public ReadOperations readOperations()
     {
+        if( !transaction.mode().allowsReads() )
```

Comment:
That is an option. The 'if' statement would occur elsewhere, so this is just moving code around a bit. We went for this approach here partly because it was more similar to where we think things will go in future when we write richer access control into the statement. Right now there is a 1:1:1 mapping from the kernel statement methods, access mode and exceptions thrown, so either solution is fine, but in future we think it might be more complex. For example, the AccessMode might not have enough knowledge to compose the message in the exception, and that would need to be done in the statement. I'd prefer to keep things the way they are here.